filebeat/input/filestream/input.go

```diff
@@ -64,5 +128,240 @@ func (inp *filestream) Run(
+
+	// TODO: NewLineReader uses additional buffering to deal with encoding and testing
+	//       for new lines in input stream. Simple 8-bit based encodings, or plain
+	//       don't require 'complicated' logic.
+	logReader, err := newFileReader(log, canceler, f, inp.readerConfig, inp.closerConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	dbgReader, err := debug.AppendReaders(logReader)
+	if err != nil {
+		f.Close()
+		return nil, err
+	}
+
+	// Configure MaxBytes limit for EncodeReader as multiplied by 4
+	// for the worst case scenario where incoming UTF32 charchers are decoded to the single byte UTF-8 characters.
+	// This limit serves primarily to avoid memory bload or potential OOM with expectedly long lines in the file.
+	// The further size limiting is performed by LimitReader at the end of the readers pipeline as needed.
+	encReaderMaxBytes := inp.maxBytes * 4
+
+	var r reader.Reader
+	r, err = readfile.NewEncodeReader(dbgReader, readfile.Config{
+		Codec:      inp.encoding,
+		BufferSize: inp.bufferSize,
+		Terminator: inp.lineTerminator,
+		MaxBytes:   encReaderMaxBytes,
+	})
+	if err != nil {
+		f.Close()
+		return nil, err
+	}
+
+	r = readfile.NewStripNewline(r, inp.lineTerminator)
+	r = readfile.NewLimitReader(r, inp.maxBytes)
+
+	return r, nil
+}
+
+// openFile opens a file and checks for the encoding. In case the encoding cannot be detected
+// or the file cannot be opened because for example of failing read permissions, an error
+// is returned and the harvester is closed. The file will be picked up again the next time
+// the file system is scanned
+func (inp *filestream) openFile(path string, offset int64) (*os.File, error) {
+	err := inp.checkFileBeforeOpening(path)
+	if err != nil {
+		return nil, err
+	}
+
+	f, err := os.OpenFile(path, os.O_RDONLY, os.FileMode(0))
+	if err != nil {
+		return nil, fmt.Errorf("failed opening %s: %s", path, err)
+	}
+
+	err = inp.initFileOffset(f, offset)
+	if err != nil {
+		f.Close()
+		return nil, err
+	}
+
+	inp.encoding, err = inp.encodingFactory(f)
+	if err != nil {
+		f.Close()
+		if err == transform.ErrShortSrc {
+			return nil, fmt.Errorf("initialising encoding for '%v' failed due to file being too short", f)
+		}
+		return nil, fmt.Errorf("initialising encoding for '%v' failed: %v", f, err)
+	}
+
+	return f, nil
+}
+
+func (inp *filestream) checkFileBeforeOpening(path string) error {
+	fi, err := os.Stat(path)
+	if err != nil {
+		return fmt.Errorf("failed to stat source file %s: %v", path, err)
+	}
+
+	if !fi.Mode().IsRegular() {
+		return fmt.Errorf("tried to open non regular file: %q %s", fi.Mode(), fi.Name())
+	}
+
+	if fi.Mode()&os.ModeNamedPipe != 0 {
+		return fmt.Errorf("failed to open file %s, named pipes are not supported", path)
+	}
+
+	return nil
+}
+
+func (inp *filestream) initFileOffset(file *os.File, offset int64) error {
+	if offset > 0 {
+		_, err := file.Seek(offset, os.SEEK_SET)
+		return err
+	}
+
+	_, err := file.Seek(0, os.SEEK_CUR)
```

Comment:
This second `Seek` looks like a no-op. Is it testing whether the given file can support seeking in general? (If so, why do we need to know that here?) If this check is necessary please add an explanatory comment

Reply:
TBH it is part of the original code in the `log` input. It does not make sense to have it here, but I was afraid to remove it. :D

Reply:
Added comment.